Use esbuild/dart-sass instead of webpack

mveytsman opened the pull request:
@josevalim I said I'd do this ages ago and then I forgot to actually push it to GitHub :(

I used postcss plugins to do the sass stuff -- we could also use https://github.com/CargoSense/dart_sass instead but then we'd need to rely on esbuild to minify the css, and still be without autoprefixer.

I think postcss is the right way until we get the esbuild equivalent for css :)

josevalim commented on config/config.exs:
```diff
@@ -5,3 +5,11 @@ config :phoenix, :stacktrace_depth, 20
 config :logger, :console, format: "[$level] $message\n"
+
+config :esbuild,
+  version: "0.12.15",
```
```suggestion
  version: "0.12.17",
```

mcrumm commented:
@mveytsman I get the autoprefixer bit, but why do we need esbuild to minify the CSS?

mveytsman commented:
@mcrumm does dart sass do css minification for us? I couldn't find it in the docs but admittedly did not look that hard.

mcrumm commented:
It does, I just wanted to make sure there wasn't something else I was missing :)

josevalim commented on assets/css/app.scss:
```diff
@@ -1,6 +1,8 @@
 @import "./app/fonts";
-@import "~bootstrap/scss/bootstrap";
+@import "../node_modules/bootstrap/scss/bootstrap.scss";
+@import  "../node_modules/uplot/dist/uPlot.min.css";
```
```suggestion
@import "../node_modules/uplot/dist/uPlot.min.css";
```

josevalim commented on config/config.exs:
```diff
@@ -5,3 +5,11 @@ config :phoenix, :stacktrace_depth, 20
 config :logger, level: :warn
 config :logger, :console, format: "[$level] $message\n"
+
+config :esbuild,
```
We need to wrap this on `if Mix.env() == :dev`.

josevalim commented on mix.exs:
```diff
@@ -55,15 +58,16 @@ defmodule Phoenix.LiveDashboard.MixProject do
       [path: path]
     else
-      [github: "phoenixframework/phoenix_live_view"]
+      [github: "phoenixframework/phoenix_live_view", branch: "mv-esbuild"]
```
This change is no longer necessary!

josevalim commented on mix.exs:
```diff
@@ -44,6 +45,8 @@ defmodule Phoenix.LiveDashboard.MixProject do
     [
       # Actual deps
       {:phoenix_live_view, "~> 0.16.0-dev", phoenix_live_view_opts()},
+      {:phoenix, "1.6.0-dev", github: "phoenixframework/phoenix", override: true},
+      {:phoenix_html, "3.0.0-dev", github: "phoenixframework/phoenix_html", override: true},
```
Rebase and we can remove both those lines now. :)

josevalim replied:
To be clear, Phoenix v1.5.10 was released with the support for function watchers.

mcrumm commented:
@mveytsman This is looking great! I do wonder how much we really need autoprefixer tho. Are we relying on it anywhere for bleeding-edge browser functionality?

Personally I'd love to see the Node requirement for LiveDashboard disappear completely :)

mveytsman commented:
@josevalim I've addressed your comments

@mcrumm I've followed your suggestion and got rid of all node in https://github.com/phoenixframework/phoenix_live_dashboard/pull/316/commits/5ce0b7681452db22c51461c5a6df57ff9b48bbe0

I vendored in bootstrap, uplot, and topbar to do this. (Note I replaced nprogress with topbar to follow the pattern in LiveView).

The only thing I don't like about this is the font - we used a webpack loader to make the woff2 file into a data url. I couldn't replicate this with dart-sass and ended up loading the base64 into a sass variable. I don't like this because it means the base64ed content isn't in sync if we ever change the font, but I think we can live with this, maybe with a comment? (Could also load the font via some cdn with a fallback if we wish!)

@mcrumm I vendored in the version of uplot that was in package.json, but it's about a year out of date - let me know if you want to bump it.

josevalim commented on assets/css/app.scss:
```diff
@@ -1,6 +1,8 @@
 @import "./app/fonts";
-@import "~bootstrap/scss/bootstrap";
+@import "../vendor/bootstrap/bootstrap.scss";
+@import  "../vendor/uPlot.min.scss";
```
```suggestion
@import "../vendor/uPlot.min.scss";
```

josevalim commented:
> maybe with a comment? (Could also load the font via some cdn with a fallback if we wish!)

A comment looks good enough to me.

The only other note is that we have some tests and we were using npm to run them. If that's important, then we most likely should keep some parts of npm (which means we no longer need to vendor the deps). Thoughts @mcrumm and @mveytsman?

mcrumm commented:
In retrospect I am fine keeping with npm around for the test/asset deps and further I am also fine with using PostCSS if that looks like the better way to go. :) Sorry for the extra work @mveytsman!

mveytsman commented:
ooof sorry I missed the tests!

I think that node-sass worked better than postcss as I had to add a bunch of plugins to get postcss to do sass without webpack.

I'll add back in node for tests and un-vendor the deps!

mcrumm commented:
Closing in favor of #382– thank you for all of your hard work on this, @mveytsman!
